### PR TITLE
Contacts

### PR DESCRIPTION
**Pull request by or-else**
Message View looks nicer: title with icon is added, unnecessary avatars removed, avatars are shown sometimes (need to figure out why not always), date formatted. 

Some static methods converted to extensions. 

Android-style toast is shown on failed login

Topic.topicTypeByName moved to Tinode.topicTypeByName due to issues with generics in swift.

### CODE REVIEW

File: TinodeSDK/model/VCard.swift [hunk 1]
```diff
@@ -24,6 +31,15 @@ public class Photo: Codable {
     convenience public init(image: UIImage) {
         self.init(type: "image/png", data: image.pngData())
     }
+
+    public func image() -> UIImage? {
```
Comment (aforge) [inline TinodeSDK/model/VCard.swift]: Very well!

File: TinodeSDK/Tinode.swift [hunk 2]
```diff
@@ -19,6 +19,21 @@ public enum TinodeError: Error {
     case notConnected(String)
     case serverResponseError(Int, String, String?)
     case notSubscribed(String)
+
+    public func description() -> String {
```
Comment (aforge) [inline TinodeSDK/Tinode.swift]: Perhaps, better to make it a property.
Function = actual work.
Property - everything else.

File: Tinodios/UiUtils.swift [hunk 3]
```diff
@@ -70,6 +71,39 @@ class UiUtils {
     }
 }
 
+extension UIViewController {
+    // Displays Android-style toast
+    func showToast(message: String, duration: TimeInterval = 3.0) {
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/UiUtils.swift]: Alerts require user interaction. They don't seem to be a good fit for cases where the user is just being informed and no action is needed. Asking the user to press a button in response to "Login failed" would seem like an overkill.

Here is an example of more iOS-e toast:

https://i.pinimg.com/originals/f1/33/f6/f133f6344775f1fb96428cb3875a6373.png

It's quite easy to do too.

File: Tinodios/MessageViewController.swift [hunk 4]
```diff
@@ -132,29 +166,50 @@ extension MessageViewController: MessagesDataSource {
...
     }
 }
 
 extension MessageViewController: MessagesDisplayDelegate, MessagesLayoutDelegate {
+
+    func configureAvatarView(_ avatarView: AvatarView, for message: MessageType, at indexPath: IndexPath, in messagesCollectionView: MessagesCollectionView) {
+        // Hide current user's avatar as well as peer's avatar in p2p topics.
+        // Avatars are useful in group topics only
+        if !isGroupTopic() || isFromCurrentSender(message: message) {
+            avatarView.isHidden = true
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/MessageViewController.swift]: I'm not sure. MessageKit suggests to do both. Maybe resources are still used for avatar even if its size is zero.

File: Tinodios/UiUtils.swift [hunk 5]
```diff
@@ -70,6 +71,39 @@ class UiUtils {
...
+        // Prevent very short toasts
+        guard duration > 0.5 else { return }
+
+        let toastLabel = UILabel()
+        toastLabel.backgroundColor = UIColor.black.withAlphaComponent(0.6)
+        toastLabel.textColor = UIColor.white
+        toastLabel.textAlignment = .center;
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/UiUtils.swift]: done

File: Tinodios/UiUtils.swift [hunk 6]
```diff
@@ -70,6 +71,39 @@ class UiUtils {
...
+        toastLabel.font = UIFont.preferredFont(forTextStyle: .footnote)
+        toastLabel.text = message
+        toastLabel.alpha = 1.0
+        toastLabel.layer.cornerRadius = 10;
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/UiUtils.swift]: done

File: Tinodios/MessageViewController.swift [hunk 7]
```diff
@@ -59,13 +67,22 @@ class MessageViewController: MessageKit.MessagesViewController, MessageDisplayLo
...
         refreshControl.addTarget(self, action: #selector(loadNextPage), for: .valueChanged)
     }
+
     override func viewDidLoad() {
         super.viewDidLoad()
-        
+
+        if let layout = messagesCollectionView.collectionViewLayout as? MessagesCollectionViewFlowLayout {
+            layout.setMessageOutgoingAvatarSize(.zero)
+            if topicType == TopicType.p2p {
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/MessageViewController.swift]: done + topic type checking added to the SDK

File: Tinodios/MessageViewController.swift [hunk 8]
```diff
@@ -80,9 +97,14 @@ class MessageViewController: MessageKit.MessagesViewController, MessageDisplayLo
         self.interactor?.cleanup()
         self.noteTimer?.invalidate()
     }
+
     @objc func loadNextPage() {
         self.interactor?.loadNextPage()
     }
+
+    private func isGroupTopic() -> Bool {
+        return topicType == TopicType.grp
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/MessageViewController.swift]: done + move d to the SDK

File: Tinodios/UiUtils.swift [hunk 9]
```diff
@@ -133,3 +167,151 @@ extension UIImage {
...
+        guard let id = id else {
+            return (UIColor.white, UIColor.gray)
+        }
+
+        let defaultBackgroundColorLight = UIColor(red: 0x9e/255, green: 0x9e/255, blue: 0x9e/255, alpha: 1.0)
+        let defaultBackgroundColorDark = UIColor(red: 0x75/255, green: 0x75/255, blue: 0x75/255, alpha: 1.0)
+        let foregroundColorDark = UIColor(red: 0xDE/255, green: 0xDE/255, blue: 0xDE/255, alpha: 1.0)
+        let foregroundColorLight = UIColor.white
+
+        let lightColors: [UIColor] = [
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/UiUtils.swift]: done

File: Tinodios/UiUtils.swift [hunk 10]
```diff
@@ -133,3 +167,151 @@ extension UIImage {
...
+            UIColor(red: 0xc5/255, green: 0xe1/255, blue: 0xa5/255, alpha: 1.0),
+            UIColor(red: 0xff/255, green: 0xf5/255, blue: 0x9d/255, alpha: 1.0),
+            UIColor(red: 0xf4/255, green: 0x8f/255, blue: 0xb1/255, alpha: 1.0),
+            UIColor(red: 0x9f/255, green: 0xa8/255, blue: 0xda/255, alpha: 1.0),
+            UIColor(red: 0xff/255, green: 0xe0/255, blue: 0x82/255, alpha: 1.0),
+            UIColor(red: 0xbc/255, green: 0xaa/255, blue: 0xa4/255, alpha: 1.0),
+            UIColor(red: 0x80/255, green: 0xde/255, blue: 0xea/255, alpha: 1.0),
+            UIColor(red: 0xce/255, green: 0x93/255, blue: 0xd8/255, alpha: 1.0)
+        ]
+        let darkColors: [UIColor] = [
```
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/UiUtils.swift]: done

File: TinodeSDK/model/VCard.swift (hunk 1 above)
Comment (or-else) [reply TinodeSDK/model/VCard.swift]: :)

File: TinodeSDK/Tinode.swift (hunk 2 above)
Comment (or-else) [reply TinodeSDK/Tinode.swift]: done

File: Tinodios/UiUtils.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (or-else) [reply Tinodios/UiUtils.swift]: here is another example: https://dribbble.com/shots/1459728--GIF-Toasts

[the remark this answers is not in the document]
Comment (aforge) [reply Tinodios/UiUtils.swift]: Okay, sounds good. Let's move forward with it.
There's plenty of other places where we need to display error messages in the app.

**Pull request merged**